components/sup/src/event/types.rs

```diff
@@ -20,33 +20,54 @@
 //! less unnecessary allocations and copying in order to create
 //! events.
 
-// TODO: It is NOT AT ALL clear that every individual event should be
-// its own struct, though that obviously works. This is just a
-// convenient and flexible (if perhaps a bit verbose) way to start.
-
-use super::Event;
-use habitat_core::{package::PackageIdent,
-                   service::ServiceGroup};
-
-#[derive(Debug)]
-pub struct ServiceStarted<'a> {
-    /// The fully-qualified identifier of the service that has just started
-    // TODO (CM): I *really* want a FullyQualifiedIdentifier type now
-    pub ident: &'a PackageIdent,
-    /// The identifier the service was loaded as. This could be (and
-    /// often is) an abbreviated identifier, like "core/redis".
-    pub spec_ident: &'a PackageIdent,
-    /// The service group of the running service
-    pub service_group: &'a ServiceGroup,
+use super::EventCore;
+use crate::{error::Result,
+            manager::service::Service};
+use prost::Message;
+
+include!(concat!(env!("OUT_DIR"), "/supervisor.event.rs"));
+
+impl Service {
+    /// Create a protobuf metadata struct for Service-related event messages.
+    pub(super) fn to_service_metadata(&self) -> ServiceMetadata {
+        ServiceMetadata { package_ident: Some(self.pkg.ident.to_string()),
+                          spec_ident:    Some(self.spec_ident.to_string()),
+                          service_group: Some(self.service_group.to_string()), }
+    }
+}
+
+impl EventCore {
+    /// Create a protobuf metadata struct for all event messages.
+    pub(super) fn to_supervisor_metadata(&self) -> SupervisorMetadata {
+        SupervisorMetadata { id: Some(self.supervisor_id.clone()), }
+    }
 }
-impl<'a> Event for ServiceStarted<'a> {}
-
-#[derive(Debug)]
-pub struct ServiceStopped<'a> {
-    /// The fully-qualified identifier of the service that has just started
-    pub ident: &'a PackageIdent,
-    //    pub spec_ident: &'a PackageIdent,
-    /// The service group of the running service
-    pub service_group: &'a ServiceGroup,
+
+pub trait EventMessage: Message + Sized {
+    /// All messages will have some top-level metadata about the
+    /// Supervisor they come from. This function allows us to set it
+    /// generically when we send the message out.
+    fn supervisor_metadata(&mut self, sup_met: SupervisorMetadata);
+
+    /// Convert a message to bytes for sending to NATS.
+    fn to_bytes(&self) -> Result<Vec<u8>> {
+        let mut buf = bytes::BytesMut::with_capacity(self.encoded_len());
+        self.encode(&mut buf)?;
+        Ok(buf.to_vec())
+    }
+}
+
+// TODO (CM): these are repetitive, and will only get more so as we
+// add events. Consider implementing via macro instead.
```

Comment:
I wonder if there's some way to leverage Prost here.

Reply:
🤔 I'll investigate.

Reply:
I haven't found a decent way to do this directly in Prost. There was some prior art in this repository for generating additional code, but it was pretty obscure, very non-obvious, and also involved lots of bare string manipulation. I'd rather not go down that path again.

I think it would be possible to create a custom derive macro for this, and then add _that_ to the `prost-build` invocation, though. However, with current Rust, that appears to require an entirely new crate for the trait in question. While that path appeals to my inner nerd (and I may investigate it further, if only for self-edification), it feels like it'd be a lot of extra ceremony for what we're doing here. 

Given that, I think for now a straightforward declarative macro will end up serving us better. That would end up being less code, more readily understandable, and more localized in scope.